File: 76.最小覆盖子串.py
#!/usr/bin/env python
# coding=utf-8
'''
Description: 
Author: yangyuxiang
Date: 2021-04-22 23:07:22
LastEditors: yangyuxiang
LastEditTime: 2021-04-22 23:35:52
FilePath: /leetcode/76.最小覆盖子串.py
'''
#
# @lc app=leetcode.cn id=76 lang=python
#
# [76] 最小覆盖子串
#

# @lc code=start
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        left = right = valid = 0
        need = {}
        window = {}
        for c in t:
            if c not in need:
                need[c] = 1
            else:
                need[c] += 1
        start = 0
        length = float("inf")
        while right < len(s):
            c = s[right]
            if c in need:
                if c not in window:
                    window[c] = 1
                else:
                    window[c] += 1
                if window[c] == need[c]:
                    valid += 1
            right += 1 
            
            while valid == len(need):
                if right - left < length:
                    start = left
                    length = right - left

                c = s[left]
                left += 1
                if c in window:
                    if window[c] == need[c]:
                        valid -= 1
                    window[c] -= 1

        return s[start:start + length] if length != float("inf") else ''


if __name__ == "__main__":
    s = 'ADOBECODEBANC'
    t = 'ABC'
    print(Solution().minWindow(s, t))


        

# @lc code=end

File: 76.最小覆盖子串.java
import java.util.HashMap;
import java.util.Map;

/*
 * @Description: 
 * @Author: yangyuxiang
 * @Date: 2021-04-23 07:39:38
 * @LastEditors: yangyuxiang
 * @LastEditTime: 2021-04-23 08:56:14
 * @FilePath: /leetcode/76.最小覆盖子串.java
 */
/*
 * @lc app=leetcode.cn id=76 lang=java
 *
 * [76] 最小覆盖子串
 */

// @lc code=start
class MinWindow {
    public String minWindow(String s, String t) {
        if(s==null||s.isEmpty()||t==null||t.isEmpty()) return "";
        int left = 0;
        int right = 0;
        Map<Character, Integer> needed = new HashMap<Character, Integer>();
        for (char c : t.toCharArray()) {
            needed.put(c, needed.getOrDefault(c, 0) + 1);
        } 
        Map<Character, Integer> window = new HashMap<Character, Integer>();
        int valid = 0;
        int start = 0;
        int len = Integer.MAX_VALUE;
        String res = "";
        
        while (right < s.length()) {
            char c = s.charAt(right);
            if (needed.containsKey(c)) {
                window.put(c, window.getOrDefault(c, 0) + 1);
                if (needed.get(c).equals(window.get(c))) {
                    valid++;
                }
            }
            right++;

            while (valid == needed.size()) {
                if (right - left < len) {
                    len = right - left;
                    res = s.substring(left, right);
                }
                char d = s.charAt(left);
                left++;
                if (needed.containsKey(d)){
                    if (needed.get(d).equals(window.get(d))) {
                        valid--;
                    }
                    window.put(d, window.getOrDefault(d, 0) - 1);
                }
            }
        }
        return res;

        
    }
}
// @lc code=end

